src/SMPCoroutines.java

```diff
@@ -0,0 +1,217 @@
+            this.id = id;
+            this.preferences = preferences;
+            acceptedMan = -1;
+            sem = new Semaphore(0);
+        }
+
+        public void setMenList(Man[] menList) {
+            this.menList = menList;
+        }
+
+        public void receiveProposal(int proposingMan) {
+            synchronized (sem) {
+                proposedMan = proposingMan;
+                sem.release();
+            }
+        }
+
+        @Override
+        public void run() {
+            while(true) {
+                try {
+                    sem.acquire();
+                    if (acceptedMan < 0) {
+                        acceptedMan = proposedMan;
+                        System.out.printf("Woman %d accepted man %d\n", id, acceptedMan);
+                    } else {
+                        int rankAccepted = -1;
+                        int rankProposed = -1;
+                        for (int i = 0; i < preferences.length; i++) {
+                            int manIndex = preferences[i] - 1;
+                            if (manIndex == proposedMan) {
+                                rankProposed = i;
+                            }
+
+                            if (manIndex == acceptedMan) {
+                                rankAccepted = i;
+                            }
+                        }
+
+                        int manToWakeUp;
+                        if (rankProposed < rankAccepted) {
+                            manToWakeUp = acceptedMan;
+                            acceptedMan = proposedMan;
+                            System.out.printf("Woman %d accepted man %d, ditching man %d\n", id, proposedMan, manToWakeUp);
+                        } else {
+                            manToWakeUp = proposedMan;
+                            System.out.printf("Woman %d rejected man %d, she is already with man %d\n", id, proposedMan, acceptedMan);
+                        }
+
+                        synchronized (menList[manToWakeUp]) {
+                            menList[manToWakeUp].notifyAll();
+                        }
+                    }
+
+
+                } catch (InterruptedException e) {
+                }
+            }
+        }
+    }
+
+
+    private int[][] prefGenderA;
+    private int[][] prefGenderB;
+
+    public SMPCoroutines(int[][] prefGenderA, int[][] prefGenderB) {
+        this.prefGenderA = prefGenderA;
+        this.prefGenderB = prefGenderB;
+    }
+
+    public String run() {
+        int count = prefGenderA.length;
+        Man[] menList = new Man[count];
+        Woman[] womenList = new Woman[count];
+
+        for (int i = 0; i < count; i++) {
+            menList[i] = new Man(i, prefGenderA[i]);
+            menList[i].setWomenList(womenList);
+
+            womenList[i] = new Woman(i, prefGenderB[i]);
+            womenList[i].setMenList(menList);
+        }
+
+
+
+        ExecutorService pool = Executors.newFixedThreadPool(count * 2);
```

Comment:
The idea of this algorithm is to create as many threads as there are participants in the matching algorithm.